documentation/modules/cruise-control/con-cruise-control-overview.adoc

```diff
@@ -2,54 +2,505 @@
+
+== Optimization goals
+
+Optimization goals define objectives for rebalancing, such as distributing topic replicas evenly across brokers. 
+
+They are categorized as follows:
+
+* *Hard goals* are preset and must be satisfied for a proposal to succeed.
+* *Soft goals* are objectives that are prioritized during optimization as much as possible, without preventing a proposal from being created if all hard goals are satisfied.
+* *Main goals* contain the list of goals that can be used by Cruise Control operations (all built-in goals by default).
+This list can be customized to specify a subset of all available goals, including any custom goals.
+* *Default goals* refer to the goals used by default when generating proposals. 
+They match the main goals unless specifically set by the user.
+* *User-provided goals* are a subset of default goals configured for specific proposals.
+
+Configure optimization goals in the `Kafka` and `KafkaRebalance` custom resources.
+
+* `Kafka` resource for hard, main, and default goals.
+** Hard goals: `Kafka.spec.cruiseControl.config.hard.goals`
+** Main goals: `Kafka.spec.cruiseControl.config.goals`
+** Default goals: `Kafka.spec.cruiseControl.config.default.goals`
+* `KafkaRebalance` resource for user-provided goals.
+** User-Provided goals: `KafkaRebalance.spec.goals`
+
+=== Hard and soft goals
+
+Hard goals are mandatory and must be satisfied for optimization proposals to be generated. 
+Soft goals are best-effort objectives that Cruise Control tries to meet after all hard goals are satisfied.
+The classification of hard and soft goals is fixed in Cruise Control code and cannot be changed.
+
+Cruise Control first prioritizes satisfying hard goals, and then addresses soft goals in the order they are listed.
+A proposal meeting all hard goals is valid, even if it violates some soft goals. 
+
+For example, a soft goal might be to evenly distribute a topic's replicas. 
+Cruise Control will ignore this if it conflicts with hard goals.
+
+Configure hard goals in your Cruise Control deployment using `Kafka.spec.cruiseControl.config.hard.goals`:
+
+* To enforce all hard goals, omit the `hard.goals` property.
+* To specify hard goals, list them in `hard.goals`.
+* To exclude a hard goal, ensure it's not in either `default.goals` or `hard.goals`.
+
+Increasing the number of configured hard goals will reduce the likelihood of Cruise Control generating optimization proposals.
+
+=== Main goals
+
+Main goals are predefined and available to all users in Cruise Control. 
+Goals not listed as main goals cannot be used in Cruise Control operations.
+Some main goals are preset as hard goals.
+
+To simplify configuration, use the inherited main goals unless you need to exclude specific goals from `KafkaRebalance` resources. 
+You can adjust the priority order in the default optimization goals configuration.
```

Comment:
To my understanding yes, the goals are listed in order of priority, so the first goal listed in the configuration will have higher priority than the second goal list

Reply:
Ok. So I think we should clarify here (and everywhere goals priority is mentioned) that the priority is based on the order. Not sure if "priority order" is used for this purpose but it's not clear to me.

Reply:
We have to admit that CC naming doesn't help much here.

IMO, this is the default inter-broker goals priority list:

https://github.com/linkedin/cruise-control/blob/main/cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/config/constants/AnalyzerConfig.java#L312-L327

To change it, the user can configure Kafka `default.goals` or KafkaRebalance `goals` with a different order. Only goals listed in Kafka `goals` can be used in the previous configuration (DEFAULT_DEFAULT_GOALS is a subset of DEFAULT_GOALS).

- Do you agree?
- Are we exposing Kafka `goals` because we intend to eventually support custom goals?